bpf: bpf host routing for tunneling

borkmann opened the pull request:
See commit msg.

borkmann committed:
bpf: enable bpf host routing for tunnels

Lift this constraint now that it is working for tunnels, too. We also
transparently get the local Pod->Pod optimization through this.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: fix up pkt for bpf host routing in tunneling mode

When switching netns when coming from overlay the packet type is not
set to HOST, so we need to do it here in order to avoid being dropped
in IP layer.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
bpf: do not blindly push to stack for bpf host routing on encap

Also in case of vxlan/geneve, let the bpf_host perform local delivery into
Pods, for example, for the case of K8s services where traffic arrives on
the phys dev and does not go via vxlan/geneve dev. For this scenario, the
same optimizations can be performed as with the direct routing case. Hence
lift the skip_redirect constraints for encaps given the return path in
bpf_lxc will also support this. Typical case for this is cloud LB pushing
inbound traffic to a node's NodePort service as one example where this will
improve performance.

For the bpf_host prog attached to the phy dev, this means that we perform
the ipv4_local_delivery() into a local Pod backend for a service more
efficiently compared to before where it gets pushed up the stack, then
routed into cilium_host and pushed from there. Note that in tunnel mode
the Pod's host-facing lxc devices do not have a policy tc egress program
attached, so the tail call into the v{4,6} policy prog of the bpf_lxc is
now done at an earlier point which also becomes visible in the RR numbers.
Only if there's no local endpoint for the target address, we push up the
stack via CTX_ACT_OK as before.

Before:

  root@apoc:~# netperf -H 192.168.180.28 -t TCP_RR -l20 -- -P 13000,12866
  MIGRATED TCP REQUEST/RESPONSE TEST from 0.0.0.0 (0.0.0.0) port 13000 AF_INET to 192.168.180.28 () port 12866 AF_INET : demo : first burst 0
  Local /Remote
  Socket Size   Request  Resp.   Elapsed  Trans.
  Send   Recv   Size     Size    Time     Rate
  bytes  Bytes  bytes    bytes   secs.    per sec

  16384  131072 1        1       20.00    8709.14
  16384  131072

After:

  root@apoc:~# netperf -H 192.168.180.28 -t TCP_RR -l20 -- -P 13000,12866
  MIGRATED TCP REQUEST/RESPONSE TEST from 0.0.0.0 (0.0.0.0) port 13000 AF_INET to 192.168.180.28 () port 12866 AF_INET : demo : first burst 0
  Local /Remote
  Socket Size   Request  Resp.   Elapsed  Trans.
  Send   Recv   Size     Size    Time     Rate
  bytes  Bytes  bytes    bytes   secs.    per sec

  16384  131072 1        1       20.00    21983.21
  16384  131072

If Pod <-> Pod traffic needs to go over vxlan/geneve, the gains will be smaller
since bpf_host needs to push to upper stack for triggering bpf_overlay. We still
do the redirect_peer() from the overlay, just that the gain might be less visible
in the big picture since the path with vxlan/geneve needs to travere upper layers
like routing/netfilter.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -256,8 +256,10 @@ handle_ipv6(struct __ctx_buff *ctx, __u32 secctx, const bool from_host)
-
 #ifdef ENCAP_IFINDEX
+	else if (!from_host)
+		return CTX_ACT_OK;
```
AFAICS, we can return here regardless of whether `ENCAP_IFINDEX` is defined. In native routing mode, the subsequent steps will all be skipped.

I think it needs a small comment though, to explicitly state the assumption being made, as that may not be clear to all datapath contributors.

borkmann replied:
Presumably the DROP_UNROUTABLE might still be relevant on the native routing case?

pchaigno replied:
Yes, but it's guarded by an `if (from_host && ...`.

borkmann replied:
Sleeping over it, I agree with your judgement, all the remainder is really only dedicated for from_host case... I'll move it out of 
 ENCAP_IFINDEX.

pchaigno commented on bpf/bpf_host.c:
```diff
@@ -522,8 +524,10 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
-
 #ifdef ENCAP_IFINDEX
+	else if (!from_host)
+		return CTX_ACT_OK;
```
This early exit will lead us to skip the `redirect()` below (line 558). That redirect was introduced by @aanm in https://github.com/cilium/cilium/commit/f54f15d289143b01d020336a2ecb0c4572df0e55 and I'm unable to say if we can actually skip it without issues. If we can, should we also skip it in the native routing case as for IPv6?

Note that `redirect()` doesn't exist in the IPv6 path.

borkmann replied:
We should probably remove that given this only relates to flannel support. Do we still actually support or test that @aanm? If its of concern, I can also disable bpf host routing in case of flannel.

borkmann replied:
I added a patch to disable bpf host routing for flannel which should address this here.

brb replied:
I think it might be not enough just disable the BPF routing when the flannel mode is detected, as https://github.com/cilium/cilium/commit/f54f15d289143b01d020336a2ecb0c4572df0e55 might be required by the CNI chaining (!= the flannel mode). @aanm Could you confirm that?

brb replied:
Ah, from the quick grepping it seems that the commit message is a bit misleading, and `HOST_REDIRECT_TO_INGRESS` is set only when running in the flannel mode. So, ignore my comment above.

pchaigno reviewed (approved):
The second commit indirectly impacts policy enforcement because whether or not we call `ipv4_local_delivery` determines where the source identity for the ingress policy enforcement is resolved (if called, then source identity is resolved in `bpf_host` and we may hit issues such as https://github.com/cilium/cilium/issues/12205). So we should make sure that we have some test coverage of policy enforcement with BPF host routing enabled.

The commits look good to me, but I'm less familiar with the L7 proxy paths and how the second commit may impact that. Maybe @joestringer can take a look?

borkmann committed:
bpf: disable bpf host routing for flannel chaining

When Cilium's datapath is chained in any way, all bets are off. Lets not
bother for such niche case for bpf host routing. Based on recent issues
(#15095, #15170) it seems like users might still run with flannel.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -295,8 +295,9 @@ func initKubeProxyReplacementOptions() (strict bool) {
-		case option.Config.Tunnel != option.TunnelDisabled:
-			msg = fmt.Sprintf("BPF host routing is only available in native routing mode.")
+		// If we chain CNIs then all bets are off.
+		case option.Config.IsFlannelMasterDeviceSet():
```
JFYI, we have deprecated the Flannel mode, and in the nearest future we should remove once we have validated that    the CNI chaining with Flannel works.

borkmann replied:
Thanks for clarifying, I wasn't quite sure. Removing the flannel specific code would actually simplify the datapath a bit. Maybe I'll make a new PR and remove flannel bits.

pchaigno replied:
Maybe we could add that to the list of guides to validate for 1.10 and remove early in the 1.11 cycle?

borkmann committed:
bpf: generally return after endpoint lookup when !from_host

After the endpoint lookup, we should generally punt up to the stack when
traffic arrives on phy dev from external (!from_host). The remainder of
the handle_ipv{4,6}() code really only deals with the case when traffic
was egressing from cilium_host device. Note that the tunnel encap handling
for the nodeport case is done elsewhere in tail_nodeport_nat_ipv{4,6}().

Suggested-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
